File: swim-js/swim-core-js/@swim/structure/main/Field.ts
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {AnyItem, Item} from "./Item";
import {AnyValue, Value} from "./Value";
import {Record} from "./Record";
import {AnyText} from "./Text";
import {AnyNum} from "./Num";

export type AnyField = Field
                     | {readonly $key: AnyValue, readonly $value: AnyValue}
                     | {[key: string]: AnyValue};

export abstract class Field extends Item {
  /** @hidden */
  constructor() {
    super();
  }

  /**
   * Always returns `true` because a `Field` can never be [[Absent]].
   */
  isDefined(): boolean {
    return true;
  }

  /**
   * Always returns `true` because a `Field` can be neither [[Extant]] nor
   * [[Absent]].
   */
  isDistinct(): boolean {
    return true;
  }

  /**
   * Returns the key component of this `Field`.
   */
  abstract get key(): Value;

  /**
   * Returns the value component of this `Field`.
   */
  abstract get value(): Value;

  /**
   * Sets the value of this `Field` to the new `value`, returning the old value.
   *
   * @throws `Error` if this `Field` is immutable.
   */
  abstract setValue(value: AnyValue): Value;

  /**
   * Returns a copy of this `Field` with the updated `value`.
   */
  abstract updatedValue(value: AnyValue): Field;

  /**
   * Returns the value component of this `Field`.
   */
  toValue(): Value {
    return this.value;
  }

  /**
   * Always returns `undefined` because a `Field` can't be a `Record`, so it
   * can't have a first member `Attr` whose key string could be returned.
   */
  tag(): string | undefined {
    return void 0;
  }

  /**
   * Always returns the value component of this `Field`.
   */
  target(): Value {
    return this.value;
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be flattened into a
   * `Value`.
   */
  flattened(): Value {
    return Item.Value.absent();
  }

  /**
   * Returns a `Record` containing just this `Field`.
   */
  unflattened(): Record {
    return Item.Record.of(this);
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have a head `Attr` whose value could be returned if its key were
   * equal to the `tag`.
   */
  header(tag: string): Value {
    return Item.Value.absent();
  }

  /**
   * Always returns `undefined` because a `Field` can't be a `Record`, so it
   * can't have a head `Attr` whose value could be returned as a `Record` if
   * its key were equal to the `tag`.
   */
  headers(tag: string): Record | undefined {
    return void 0;
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have a first member.
   */
  head(): Item {
    return Item.absent();
  }

  /**
   * Always returns an empty `Record` because a `Field` can't itself be a
   * `Record`, so it can't have any non-first members.
   */
  tail(): Record {
    return Item.Record.empty();
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have any non-first members to flatten, and because a `Field` isn't
   * a distinct `Value`, so it can't return `Extant`.
   */
  body(): Value {
    return Item.Value.absent();
  }

  /**
   * Always returns `0` because a `Field` can't be a `Record`, so it can't
   * contain any members.
   */
  get length(): number {
    return 0;
  }

  /**
   * Always returns `false` because a `Field` can't be a `Record`, so it can't
   * have a `Field` member whose key is equal to the given `key`.
   */
  has(key: AnyValue): boolean {
    return false;
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have a `Field` member whose key is equal to the given `key`.
   */
  get(key: AnyValue): Value {
    return Item.Value.absent();
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have an `Attr` member whose key is equal to the given `key`.
   */
  getAttr(key: AnyText): Value {
    return Item.Value.absent();
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have a `Slot` member whose key is equal to the given `key`.
   */
  getSlot(key: AnyValue): Value {
    return Item.Value.absent();
  }

  /**
   * Always returns `undefined` because a `Field` can't be a `Record`, so it
   * can't have a `Field` member whose key is equal to the given `key`.
   */
  getField(key: AnyValue): Field | undefined {
    return void 0;
  }

  /**
   * Always returns [[Absent]] because a `Field` can't be a `Record`, so it
   * can't have a member at the given `index`.
   */
  getItem(index: AnyNum): Item {
    return Item.absent();
  }

  deleted(key: AnyValue): Field {
    return this;
  }

  conditional(thenTerm: Field, elseTerm: Field): Field;
  conditional(thenTerm: AnyItem, elseTerm: AnyItem): Item;
  conditional(thenTerm: AnyItem, elseTerm: AnyItem): Item {
    thenTerm = Item.fromAny(thenTerm);
    return thenTerm;
  }

  or(that: Field): Field;
  or(that: AnyItem): Item;
  or(that: AnyItem): Item {
    return this;
  }

  and(that: Field): Field;
  and(that: AnyItem): Item;
  and(that: AnyItem): Item {
    that = Item.fromAny(that);
    return that;
  }

  lambda(template: Value): Value {
    return Item.Value.absent();
  }

  /**
   * Converts the value of this `Field` into a `string` value, if possible.
   *
   * @throws `Error` if the value of this `Field` can't be converted into a
   *         `string` value.
   */
  stringValue(): string | undefined;
  /**
   * Converts the value of this `Field` into a `string` value, if possible;
   * otherwise returns `orElse` if the value of this `Field` can't be converted
   * into a `string` value.
   */
  stringValue<T>(orElse: T): string | T;
  stringValue<T>(orElse?: T): string | T | undefined {
    return this.value.stringValue(orElse);
  }

  /**
   * Converts the value of this `Field` into a `number` value, if possible.
   *
   * @throws `Error` if the value of this `Field` can't be converted into a
   *         `number` value.
   */
  numberValue(): number | undefined;
  /**
   * Converts the value of this `Field` into a `number` value, if possible;
   * otherwise returns `orElse` if the value of this `Field` can't be converted
   * into a `number` value.
   */
  numberValue<T>(orElse: T): number | T;
  numberValue<T>(orElse?: T): number | T | undefined {
    return this.value.numberValue(orElse);
  }

  /**
   * Converts the value of this `Field` into a `boolean` value, if possible.
   *
   * @throws `Error` if the value of this `Field` can't be converted into a
   *         `boolean` value.
   */
  booleanValue(): boolean | undefined;
  /**
   * Converts the value of this `Field` into a `boolean` value, if possible;
   * otherwise returns `orElse` if the value of this `Field` can't be converted
   * into a `boolean` value.
   */
  booleanValue<T>(orElse: T): boolean | T;
  booleanValue<T>(orElse?: T): boolean | T | undefined {
    return this.value.booleanValue(orElse);
  }

  abstract toAny(): AnyField;

  abstract branch(): Field;

  abstract clone(): Field;

  abstract commit(): this;

  /** @hidden */
  static readonly IMMUTABLE: number = 1 << 0;

  static of(key: AnyValue, value?: AnyValue): Field {
    let name;
    if (typeof key === "string") {
      name = key;
    } else if (key instanceof Item.Text) {
      name = key.value;
    }
    if (name !== void 0 && name.charCodeAt(0) === 64/*'@'*/) {
      arguments[0] = name.slice(1);
      return Item.Attr.of.apply(undefined, arguments);
    } else {
      return Item.Slot.of.apply(undefined, arguments);
    }
  }

  static fromAny(field: AnyField): Field {
    if (field instanceof Field) {
      return field;
    } else if (field && typeof field === "object") {
      if ((field as any).$key !== void 0) {
        return Field.of((field as any).$key, (field as any).$value);
      } else {
        for (const key in field) {
          return Field.of(key, (field as any)[key]);
        }
      }
    }
    throw new TypeError("" + field);
  }
}
Item.Field = Field;

File: swim-java/swim-core-java/swim.structure/src/main/java/swim/structure/Field.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.structure;

import java.math.BigInteger;
import java.util.Map;
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;

public abstract class Field extends Item implements Map.Entry<Value, Value> {
  volatile int flags;

  Field() {
    // stub
  }

  /**
   * Always returns {@code true} because a {@code Field} can never be
   * {@link Absent}.
   */
  @Override
  public final boolean isDefined() {
    return true;
  }

  /**
   * Always returns {@code true} because a {@code Field} can be neither
   * {@link Extant} nor {@link Absent}.
   */
  @Override
  public final boolean isDistinct() {
    return true;
  }

  /**
   * Returns the key component of this {@code Field}.
   */
  @Override
  public abstract Value key();

  /**
   * Returns the key component of this {@code Field}.  Equivalent to {@link
   * #key()}.
   */
  @Override
  public Value getKey() {
    return key();
  }

  /**
   * Returns the value component of this {@code Field}.
   */
  public abstract Value value();

  /**
   * Returns the value component of this {@code Field}.  Equivalent to {@link
   * #value()}.
   */
  @Override
  public Value getValue() {
    return value();
  }

  /**
   * Sets the value of this {@code Field} to the new {@code value}, returning
   * the old value.
   *
   * @throws UnsupportedOperationException if this {@code Field} is immutable.
   */
  @Override
  public abstract Value setValue(Value value);

  /**
   * Returns a copy of this {@code Field} with the updated {@code value}.
   */
  public abstract Field updatedValue(Value value);

  /**
   * Returns the value component of this {@code Field}.  Equivalent to {@link
   * #value()}.
   */
  @Override
  public Value toValue() {
    return value();
  }

  /**
   * Always returns {@code null} because a {@code Field} can't be a {@code
   * Record}, so it can't have a first member {@code Attr} whose key string
   * could be returned.
   */
  @Override
  public final String tag() {
    return null;
  }

  /**
   * Always returns the value component of this {@code Field}.
   */
  @Override
  public final Value target() {
    return value();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be flattened
   * into a {@code Value}.
   */
  @Override
  public final Value flattened() {
    return Value.absent();
  }

  /**
   * Returns a {@code Record} containing just this {@code Field}.
   */
  @Override
  public final Record unflattened() {
    return Record.of(this);
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a head {@code Attr} whose value could be
   * returned if its key were equal to the {@code tag}.
   */
  @Override
  public final Value header(String tag) {
    return Value.absent();
  }

  /**
   * Always returns {@code null} because a {@code Field} can't be a {@code
   * Record}, so it can't have a head {@code Attr} whose value could be
   * returned as a {@code Record} if its key is equal to the {@code tag}.
   */
  @Override
  public final Record headers(String tag) {
    return null;
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a first member.
   */
  @Override
  public final Item head() {
    return Item.absent();
  }

  /**
   * Always returns an empty {@code Record} because a {@code Field} can't
   * itself be a {@code Record}, so it can't have any non-first members.
   */
  @Override
  public final Record tail() {
    return Record.empty();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have any non-first members to flatten, and because a
   * {@code Field} isn't a distinct {@code Value}, so it can't return {@code
   * Extant}.
   */
  @Override
  public final Value body() {
    return Value.absent();
  }

  /**
   * Always returns {@code 0} because a {@code Field} can't be a {@code Record},
   * so it can't contain any members.
   */
  @Override
  public final int length() {
    return 0;
  }

  /**
   * Always returns {@code false} because a {@code Field} can't be a {@code
   * Record}, so it can't have a member equal to {@code item}.
   */
  @Override
  public final boolean contains(Item item) {
    return false;
  }

  /**
   * Always returns {@code false} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key is equal to the
   * given {@code key}.
   */
  @Override
  public final boolean containsKey(Value key) {
    return false;
  }

  /**
   * Always returns {@code false} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key string is equal
   * to the given {@code key}.
   */
  @Override
  public final boolean containsKey(String key) {
    return false;
  }

  /**
   * Always returns {@code false} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose value is equal to
   * the given {@code value}.
   */
  @Override
  public final boolean containsValue(Value value) {
    return false;
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key is equal to the
   * given {@code key}.
   */
  @Override
  public final Value get(Value key) {
    return Value.absent();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key string is equal
   * to the given {@code key}.
   */
  @Override
  public final Value get(String key) {
    return Value.absent();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have an {@code Attr} member whose key is equal to the
   * given {@code key}.
   */
  @Override
  public final Value getAttr(Text key) {
    return Value.absent();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have an {@code Attr} member whose key string is equal
   * to the given {@code key}.
   */
  @Override
  public final Value getAttr(String key) {
    return Value.absent();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Slot} member whose key is equal to the
   * given {@code key}.
   */
  @Override
  public final Value getSlot(Value key) {
    return Value.absent();
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Slot} member whose key string is equal
   * to the given {@code key}.
   */
  @Override
  public final Value getSlot(String key) {
    return Value.absent();
  }

  /**
   * Always returns {@code null} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key is equal to the
   * given {@code key}.
   */
  @Override
  public final Field getField(Value key) {
    return null;
  }

  /**
   * Always returns {@code null} because a {@code Field} can't be a {@code
   * Record}, so it can't have a {@code Field} member whose key string is equal
   * to the given {@code key}.
   */
  @Override
  public final Field getField(String key) {
    return null;
  }

  /**
   * Always returns {@link Absent} because a {@code Field} can't be a {@code
   * Record}, so it can't have a member at the given {@code index}.
   */
  @Override
  public final Item getItem(int index) {
    return Item.absent();
  }

  @Override
  public Field removed(Value key) {
    return this;
  }

  @Override
  public Field removed(String key) {
    return this;
  }

  @Override
  public Item conditional(Item thenTerm, Item elseTerm) {
    if (thenTerm instanceof Field && elseTerm instanceof Field) {
      return conditional((Field) thenTerm, (Field) elseTerm);
    }
    return thenTerm;
  }

  public Field conditional(Field thenTerm, Field elseTerm) {
    return thenTerm;
  }

  @Override
  public Item or(Item that) {
    if (that instanceof Field) {
      return or((Field) that);
    }
    return this;
  }

  public Field or(Field that) {
    return this;
  }

  @Override
  public Item and(Item that) {
    if (that instanceof Field) {
      return and((Field) that);
    }
    return that;
  }

  public Field and(Field that) {
    return that;
  }

  @Override
  public Value lambda(Value template) {
    return Value.absent();
  }

  /**
   * Converts the value of this {@code Field} into a {@code String} value,
   * if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a {@code String} value.
   */
  @Override
  public final String stringValue() {
    return getValue().stringValue();
  }

  /**
   * Converts the value of this {@code Field} into a {@code String} value,
   * if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a {@code string} value.
   */
  @Override
  public final String stringValue(String orElse) {
    return getValue().stringValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code byte}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code byte} value.
   */
  @Override
  public final byte byteValue() {
    return getValue().byteValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code byte}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code byte} value.
   */
  @Override
  public final byte byteValue(byte orElse) {
    return getValue().byteValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code short}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code short} value.
   */
  @Override
  public final short shortValue() {
    return getValue().shortValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code short}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code short} value.
   */
  @Override
  public final short shortValue(short orElse) {
    return getValue().shortValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code int}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code int} value.
   */
  @Override
  public final int intValue() {
    return getValue().intValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code int}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code int} value.
   */
  @Override
  public final int intValue(int orElse) {
    return getValue().intValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code long}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code long} value.
   */
  @Override
  public final long longValue() {
    return getValue().longValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code long}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code long} value.
   */
  @Override
  public final long longValue(long orElse) {
    return getValue().longValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code float}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code float} value.
   */
  @Override
  public final float floatValue() {
    return getValue().floatValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code float}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code float} value.
   */
  @Override
  public final float floatValue(float orElse) {
    return getValue().floatValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code double}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code double} value.
   */
  @Override
  public final double doubleValue() {
    return getValue().doubleValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code double}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code double} value.
   */
  @Override
  public final double doubleValue(double orElse) {
    return getValue().doubleValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a {@code BigInteger} value,
   * if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a {@code BigInteger} value.
   */
  @Override
  public final BigInteger integerValue() {
    return getValue().integerValue();
  }

  /**
   * Converts the value of this {@code Field} into a {@code BigInteger} value,
   * if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a {@code BigInteger} value.
   */
  @Override
  public final BigInteger integerValue(BigInteger orElse) {
    return getValue().integerValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a {@code Number} object,
   * if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a {@code Number} object.
   */
  @Override
  public final Number numberValue() {
    return getValue().numberValue();
  }

  /**
   * Converts the value of this {@code Field} into a {@code Number} object,
   * if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a {@code Number} object.
   */
  @Override
  public final Number numberValue(Number orElse) {
    return getValue().numberValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code char}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code char} value.
   */
  @Override
  public final char charValue() {
    return getValue().charValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code char}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code char} value.
   */
  @Override
  public final char charValue(char orElse) {
    return getValue().charValue(orElse);
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code boolean}
   * value, if possible.
   *
   * @throws UnsupportedOperationException if the value of this {@code Field}
   *                                       can't be converted into a primitive {@code boolean} value.
   */
  @Override
  public final boolean booleanValue() {
    return getValue().booleanValue();
  }

  /**
   * Converts the value of this {@code Field} into a primitive {@code boolean}
   * value, if possible; otherwise returns {@code orElse} if the value of this
   * {@code Field} can't be converted into a primitive {@code boolean} value.
   */
  @Override
  public final boolean booleanValue(boolean orElse) {
    return getValue().booleanValue(orElse);
  }

  @Override
  public abstract Field branch();

  @Override
  public abstract Field commit();

  static final int IMMUTABLE = 1 << 0;
  static final AtomicIntegerFieldUpdater<Field> FLAGS =
      AtomicIntegerFieldUpdater.newUpdater(Field.class, "flags");

  public static Field of(Object object) {
    if (object instanceof Field) {
      return (Field) object;
    } else if (object instanceof Map.Entry<?, ?>) {
      final Map.Entry<?, ?> entry = (Map.Entry<?, ?>) object;
      return Slot.of(Value.fromObject(entry.getKey()), Value.fromObject(entry.getValue()));
    } else {
      throw new IllegalArgumentException(object.toString());
    }
  }
}
